src/cargo/util/lock.rs

```diff
@@ -0,0 +1,92 @@
+use file_lock::filename::{Mode, ParseError};
+use file_lock::filename::Lock as FileLock;
+use file_lock::filename::Error as FileLockError;
+
+use std::path::PathBuf;
+use std::fs;
+use std::thread::sleep_ms;
+
+use util::{Config, CargoError, CargoResult, human, caused_human};
+
+pub use file_lock::filename::Kind as LockKind;
+
+impl From<FileLockError> for Box<CargoError> {
+    fn from(t: FileLockError) -> Self {
+        Box::new(t)
+    }
+}
+
+impl From<ParseError> for Box<CargoError> {
+    fn from(t: ParseError) -> Self {
+        Box::new(t)
+    }
+}
+
+impl CargoError for FileLockError {
+    fn is_human(&self) -> bool { true }
+}
+impl CargoError for ParseError {}
+
+pub struct CargoLock {
+    kind: LockKind, 
+    inner: FileLock,
+}
+
+impl CargoLock {
+
+    pub fn lock_kind(config: &Config) -> CargoResult<LockKind> {
+        // TODO(ST): rename config key to something more suitable
+        const CONFIG_KEY: &'static str = "build.lock-kind";
+        match try!(config.get_string(CONFIG_KEY)).map(|t| t.0) {
+            None => Ok(LockKind::NonBlocking),
+            Some(kind_string) => match kind_string.parse() {
+                Ok(kind) => Ok(kind),
+                Err(_) => Err(human(format!("Failed to parse value '{}' at configuration key \
+                                            '{}'. Must be one of '{}' and '{}'",
+                                            kind_string, CONFIG_KEY,
+                                            LockKind::NonBlocking.as_ref(), 
+                                            LockKind::Blocking.as_ref())))
+            }
+        }
+    }
+
+    pub fn new(path: PathBuf, kind: LockKind) -> CargoLock {
+        CargoLock {
+            kind: kind,
+            inner: FileLock::new(path, Mode::Write)
+        }
+    }
+
+    pub fn new_shared(path: PathBuf, kind: LockKind) -> CargoLock {
+        CargoLock {
+            kind: kind,
+            inner: FileLock::new(path, Mode::Read)
+        }
+    }
+
+    pub fn lock(&mut self) -> CargoResult<()> {
+        // NOTE(ST): This could fail if cargo is run concurrently for the first time
```

Comment:
This "NOTE(ST)" and "TODO(ST)" isn't currently present throughout the rest of Cargo, can you remove these annotations?